### PR TITLE
Propagate module-linking types to `wasmtime`

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit adds lots of plumbing to get the type section from the
module linking proposal plumbed all the way through to the `wasmtime`
crate and the `wasmtime-c-api` crate. This isn't all that useful right
now because Wasmtime doesn't support imported/exported
modules/instances, but this is all necessary groundwork to getting that
exported at some point. I've added some light tests but I suspect the
bulk of the testing will come in a future commit.

One major change in this commit is that `SignatureIndex` no longer
follows type type index space in a wasm module. Instead a new
`TypeIndex` type is used to track that. Function signatures, still
indexed by `SignatureIndex`, are then packed together tightly.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Ok I have now rebased this as well!

File: cranelift/wasm/src/translation_utils.rs [hunk 1]
```diff
@@ -73,6 +73,54 @@ entity_impl!(DataIndex);
...
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+pub enum EntityType {
+    Global(Global),
+    Memory(Memory),
+    Table(Table),
+    Function(TypeIndex),
+    Instance(TypeIndex),
+    Module(TypeIndex),
+}
```
Comment (fitzgen) [inline cranelift/wasm/src/translation_utils.rs:122]: Was the `allow(missing_docs)` intended to be temporary while getting something up and running?

File: crates/c-api/src/types/instance.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            _ => None,
+        }
+    }
+
+    // pub(crate) fn ty(&self) -> &CInstanceType {
+    //     match &self.ext.which {
+    //         CExternType::Instance(f) => &f,
+    //         _ => unsafe { std::hint::unreachable_unchecked() },
+    //     }
+    // }
```
Comment (fitzgen) [inline crates/c-api/src/types/instance.rs:38]: Commented out code?

Also, is getting the `ty` so perf sensitive that we have to use potentially-UB-inducing `unreachable_unchecked` over regular `unreachable`?

File: crates/c-api/src/types/module.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            _ => None,
+        }
+    }
+
+    // pub(crate) fn ty(&self) -> &CModuleType {
+    //     match &self.ext.which {
+    //         CExternType::Module(f) => &f,
+    //         _ => unsafe { std::hint::unreachable_unchecked() },
+    //     }
+    // }
```
Comment (fitzgen) [inline crates/c-api/src/types/module.rs:38]: ditto

**Review by fitzgen — APPROVED** [review]

File: crates/c-api/src/types/instance.rs (hunk 2 above)
Comment (fitzgen) [reply crates/c-api/src/types/instance.rs:38]: I guess this is going into an FFI language boundary... Did the auto-abort on unwind across FFI stuff re-merge? If so, then we should use `unreachable!()`, if not then we should abort ourselves I think.

File: cranelift/wasm/src/translation_utils.rs (hunk 1 above)
Comment (alexcrichton) [reply cranelift/wasm/src/translation_utils.rs:122]: Oh I figured the docs wouldn't really go much beyond what the code already says here, but I can add some strings.

File: crates/c-api/src/types/instance.rs (hunk 2 above)
Comment (alexcrichton) [reply crates/c-api/src/types/instance.rs:38]: Oh this was copied from the other types in the C API, but I think it makes more sense to just add this back in as necessary later on.

Also yeah I doubt that `unreachable_unchecked` is really worth it here, and an `unreachable!()` should be fine.

**Pull request merged**